RFC: add new var system

jsquyres opened the pull request:
Stop the proliferation of unstandardized environment variables and use a standardized convention.

The first commit adds the var system: `fi_var_register()`, `fi_var_get()`, and `fi_var_fini()`.

The second commit converts the sockets provider to use the var interface:
* SOCK_PE_WAITTIME -> FI_SOCKETS_PE_WAITTIME
* SOCK_DGRAM_DROP_RATE -> FI_SOCKETS_DGRAM_DROP_RATE

@jithinjosepkl @shantonu please review

The third commit converts the PSM provider to use the var interface (note: I cannot compile/test this commit, and I'm quite sure that my help strings are totally wrong):
* OFI_PSM_NAME_SERVER -> FI_PSM_NAME_SERVER
* OFI_PSM_AM_MSG -> FI_PSM_AM_MSG
* OFI_PSM_TAGGED_RMA -> FI_PSM_TAGGED_RMA
* OFI_PSM_UUID -> FI_PSM_UUID

@j-xiong please review

jsquyres commented:
Note that #177 is the original issue from which this PR emerged.

shefty commented:
Note that Jianxin is out until Thursday of this week.

shefty commented on prov/psm/src/psmx_init.c:
```diff
@@ -450,11 +450,25 @@ struct fi_provider psmx_prov = {
+			"Failed to register name variable\n", name);
+	if (fi_var_get(&psmx_prov, name, value))
+		FI_INFO(&psmx_prov, FI_LOG_CORE,
+			"Failed to read uuid variable\n");
```
We should be able to avoid these FI_INFO calls in each provider by moving them into the fi_var_* calls.

jsquyres replied:
Sure; done.

shefty commented on src/var.c:
```diff
@@ -0,0 +1,138 @@
+	free(str);
+
+	return FI_SUCCESS;
+}
```
Please return negative errno values in error cases.

jsquyres replied:
K.

shefty commented on src/var.c:
```diff
@@ -0,0 +1,138 @@
+			free(str);
+
+			return FI_SUCCESS;
+		}
```
nit - can you add braces to the for loop?  I get lost with multi-line code segments

jsquyres replied:
Sure.  I actually like braces even for 1-line blocks (defensive programming and all that).  But I thought you kernel guys were allergic to braces, so I left them off whenever possible.  I'm happy to add them back.  :-)

shefty commented on include/rdma/fi_var.h:
```diff
@@ -0,0 +1,85 @@
+ * registered via fi_var_register().
+ */
+int fi_var_get(struct fi_provider *provider, const char *var_name,
+	char **value);
```
Does it make sense to provide helper wrappers that convert value to the correct value?  E.g.

int fi_var_get_int(*provider, *var_name, *default_value);

jsquyres replied:
Mmm.  Good point.  I don't have an opinion here -- I'm happy to add the following:

* `_get_str()`
* `_get_int()`
* `_get_long()` 
* `_get_bool()` (which takes 0, true, yes, on, 1, false, no, off -- case insensitive, of course)

That would cover all the cases in sockets and psm.

Want me to add these?

shefty replied:
Yes, please.  Thanks

shantonu commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -580,19 +581,29 @@ SOCKETS_INI
+				"How many miliseconds to spin while waiting for progress"))
+		SOCK_LOG_INFO("Failed to register pe_waittime variable\n");
+	if (fi_var_get(&sock_prov, "pe_waittime", &value))
+		SOCK_LOG_INFO("Failed to read pe_waittime variable\n");
```
Since the var is now FI_SOCKETS_*, should the log report the same? The name of the var is not consistent in the logs.

jsquyres replied:
Fair enough -- there's actually a subtle distinction between the FI configuration variable name and how it is exposed.  Currently, FI config variables are exposed in the environment via the form $FI_(upper_case(provider_name))_(upper_case(variable_name)).  E.g.:

```
shell$ export FI_SOCKETS_PE_WAITTIME=10
```

But I would expect that if we supported config files someday, the variable name would be exposed via "(prover_name)_(variable_name)", e.g., perhaps:

```ini
sockets_pe_waittime = 10
```

(and perhaps the name in the config file would be case insensitive...?)

So there's two things different:

1. The `FI_` prefix in the env var case
1. The upper / lower case issues

I feel strongly that we need the first one (`FI_` prefix), but I could be talked out of the second (upper/lower case transmogrification).

shefty replied:
I agree with keeping FI_.  I would go with case insensitive.

shantonu commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -580,19 +581,29 @@ SOCKETS_INI
+		SOCK_LOG_INFO("Failed to read dgram_drop_rate variable\n");
+	if (value) {
+		sock_dgram_drop_rate = strtol(value, NULL, 10);
+		SOCK_LOG_INFO("SOCK_DGRAM_DROP_RATE = %d\n",
```
Same here

pmmccorm commented on src/var.c:
```diff
@@ -0,0 +1,138 @@
+				asprintf(&str, "read var %s=%s\n",
+					var_name, *value);
+			else
+				asprintf(&str, "read var %s=<not set>\n",
```
Can we avoid the malloc every iteration with asprintf by doing something like:

    FI_INFO(provider, FI_LOG_CORE, "read var %s=%s\n", var_name, value);

?

jsquyres replied:
I missed that FI_INFO was varargs.  Sure, that's a whole lot better -- I'll update here and the other location.

jsquyres commented:
I just pushed some new commits, which hid some of the prior comments by default on the PR.  I think there's two main open questions left:

1. @shefty Whether we want more `fi_var_get_TYPE()` functions (https://github.com/ofiwg/libfabric/pull/1055#discussion_r31969918)
1. @shantonu Whether we like the dichotomy of the registered variable name vs. how it appears in the environment and/or log messages (https://github.com/ofiwg/libfabric/pull/1055#discussion-diff-32030642)

shefty commented on prov/psm/src/psmx_init.c:
```diff
@@ -475,10 +485,19 @@ PSM_INI
+		psmx_get_int_env("tagged_rma",
+				"Whether PSM supports tagged RMA or not",
+				1);
+	psmx_reg_and_read("uuid", "Unique ID for PSM fabric", &psmx_env.uuid);
```
@j-xiong - This is separate from this patch, but can you look at moving reading these values until immediately before they are needed?

j-xiong replied:
Yes, will do that in a separate PR.

j-xiong replied:
"Unique Job ID required by PSM fabric"

jsquyres replied:
Fixed.

shefty commented on prov/sockets/src/sock_fabric.c:
```diff
@@ -580,20 +581,21 @@ SOCKETS_INI
+			"Drop every Nth dgram frame (debug only)");
+	fi_var_get(&sock_prov, "dgram_drop_rate", &value);
+	if (value)
+		sock_dgram_drop_rate = strtol(value, NULL, 10);
```
@jithinjosepkl - This is separate from this patch, but can you look at moving reading these values until immediately before they are needed?

jithinjosepkl replied:
Sure @shefty, makes sense.

pmmccorm commented on src/var.c:
```diff
@@ -0,0 +1,154 @@
+	v->next = fi_vars;
+	fi_vars = v;
+
+	asprintf(&str, "registered var %s\n", var_name);
```
Ditto here about asprintf.

j-xiong commented on prov/psm/src/psmx_init.c:
```diff
@@ -475,10 +485,19 @@ PSM_INI
-	psmx_env.uuid		= getenv("OFI_PSM_UUID");
+	psmx_env.name_server =
+		psmx_get_int_env("name_server",
+				"Whether there is a PSM name server or not",
```
"Whether to turn on the name server or not"

jsquyres replied:
Fixed.

j-xiong commented on prov/psm/src/psmx_init.c:
```diff
@@ -475,10 +485,19 @@ PSM_INI
+				1);
+	psmx_env.am_msg =
+		psmx_get_int_env("am_msg",
+				"Whether PSM supports active messages or not",
```
"Whether to use active message based messaging or not (default no)"

jsquyres replied:
Fixed.

j-xiong commented on prov/psm/src/psmx_init.c:
```diff
@@ -475,10 +485,19 @@ PSM_INI
+				0);
+	psmx_env.tagged_rma =
+		psmx_get_int_env("tagged_rma",
+				"Whether PSM supports tagged RMA or not",
```
"Whether to use tagged message for large size RMA or not (default yes)"

jsquyres replied:
Fixed.

jsquyres commented:
New commits pushed:

* Updated var interface to have "get" methods for string, int, long, and boolean
* Updated psm and sockets providers to match (I still can't compile/test the PSM provider, sorry!)

Comments on the new proposed API welcome (this is an RFC, after all).  See fi_var.h for a description of the API functions.

shefty commented on src/var.c:
```diff
@@ -0,0 +1,247 @@
+				*value = 1;
+				FI_INFO(provider, FI_LOG_CORE,
+					"read boolean var %s=true\n", var_name);
+				return -FI_ENODATA;
```
Should return FI_SUCCESS here.

jsquyres replied:
Yoinks; thanks.

shefty commented on src/var.c:
```diff
@@ -0,0 +1,247 @@
+			FI_INFO(provider, FI_LOG_CORE,
+				"read string var %s=<not set>\n", var_name);
+			return -FI_ENODATA;
+		}
```
You can drop the 'return FI_SUCCESS' line and just let the code fall through.  And change 'return -FI_ENODATA' to 'ret = -FI_ENODATA', which will give us a single exit point.  This same flow can be followed below.

shefty commented:
I'm pretty happy with this.

shefty commented on include/rdma/fi_var.h:
```diff
@@ -0,0 +1,119 @@
+
+/* Similar to fi_var_get_str(), but the value is converted to an
+ * long.  No checking is done to ensure that the value the user set
+ * is actually an integer -- atoi() is simply called on whatever value
```
actually uses strtol(), rather than atoi().

jsquyres commented:
New commits pushed.

Would like to get signoffs from the sockets and PSM providers before merging (I cannot test the PSM stuff).

jithinjosepkl commented:
:+1: from sockets. But, I was just wondering where we use the help strings.
Is there a way user can list all the options for a particular provider?
(May be this is something thats already in plan?)

jsquyres commented:
Ja, the plan for the help strings is to output them in fi_info, or somesuch.  Not sure yet -- but there will be some kind of CLI utility to display them for the user.
